feat: add fine-grained policy merge

eduardacoppo opened the pull request:
Previously here: https://github.com/tidewise/tools-syskit/pull/3
This was on top of another `tools/syskit` PR

eduardacoppo committed:
feat: add fine-grained policy merge

Previously, if a parameter was added to the designer-provided policy, no automated policy determination would be performed. That meant that the init flag would not be computed in this case.
To fix this, a merge_policy function was implemented to merge the explicit policy (the designer-provided one) and the computed policy (init, for instance)
The fix follows these rules:
- if a value is in policy, use it;
- otherwise use the value from computed_policy
Also, if the type policy is set to data, the size policy must be removed, as its only meaningful for the type buffer and it causes the connection to fail.

The Ruby method `merge` merges two hashes. According to the documentation:
"Returns a new hash containing the contents of other_hash and the contents of hsh. If no block is specified, the value for entries with duplicate keys will be that of other_hash."
In this case, to follow the rules of prioritizing the value from explicit_policy in case of key duplication, 'other_hash' is explicit_policy and 'hsh' is computed_policy

eduardacoppo committed:
test: add tests and update existing

jhonasiv commented on lib/syskit/network_generation/dataflow_dynamics.rb:
```diff
@@ -553,6 +553,16 @@ def compute_connection_policies
+            def merge_policy(explicit_policy, computed_policy)
+                merged_policy = computed_policy.merge(explicit_policy)
+
+                if merged_policy[:type] == :data
```
how could this happen? due to a bad explicit policy?

doudou reviewed (commented):
This is a good first step, but there's one issue.

When using `needs_reliable_connection`, we have to add period information to the syskit models. Until now, setting the policy explicitly would be enough to avoid having to do this (because of time, or because Syskit limitations would make it hard to do). This PR will break these cases.

My suggestion is to split policy_for into:
- policy_compute_data_element(policy, fallback_policy)
- policy_default_init_flag

Always call the latter, but call the former only if the explicit policy does not have `:type` set already. Create corresponding unit tests.

eduardacoppo committed:
fix: split policy_for

eduardacoppo committed:
test: add tests for new functions

doudou commented on lib/syskit/network_generation/dataflow_dynamics.rb:
```diff
@@ -618,9 +627,22 @@ def policy_for(
+                source_task, source_port_name, sink_port_name, sink_task, fallback_policy
+            )
+                policy = {}
+                unless policy[:type]
```
`policy` is the empty hash here. This will always be nil.

The `if` should obviously be checking the policy that is coming from the policy graph.

I would actually pass `policy` from the loop in `compute_policies_from` to the method, and add the merge_policy call at the end of this method instead of doing it in said loop.

doudou commented on test/network_generation/test_dataflow_dynamics.rb:
```diff
@@ -360,6 +363,151 @@ module NetworkGeneration
                     flexmock(@dynamics).should_receive(:propagate).with(tasks[0, 2])
                 end
+
+                it "handles the case where the explicit policy sets the type to :data" do
```
And what about the type being `:buffer`. We should make sure it's not overriding the size.

doudou commented on test/network_generation/test_dataflow_dynamics.rb:
```diff
@@ -360,6 +363,151 @@ module NetworkGeneration
+
+                    task0.out_port.connect_to task1.in_port, type: :data
+
+                    flexmock(@dynamics)
+                        .should_receive(:policy_for)
+                        .with(task0, "out", "in", task1, nil)
+                        .and_return(type: :buffer, size: 10, init: true)
```
This is *massively* cheating.

What we want to check is that `policy_compute_data_element` is not called when :type is set. That's the really important bit.

eduardacoppo committed:
fix: merge explicit and computed policies

renamed variables and simplified code

eduardacoppo committed:
chore: add missing init arg

eduardacoppo committed:
fix/test: update and fix the merge tests

- Updated test expectations
- Updated mock expectations
- Added test to make sure the size is not overridden when an explicit policy specifies `type: :buffer`
- Fixed the merging logic in the reliable connection policy test, properly verifying that `merge_policy` is called instead of directly modifying the expected policy
About this last one. I'm not sure if that's the best way to do it. Tried to think of a way of simplifying it, but the test expects this chain of mocks.

eduardacoppo commented on lib/syskit/network_generation/dataflow_dynamics.rb:
```diff
@@ -561,25 +571,24 @@ def compute_policies_from(connection_graph, source_task, policy_graph = {})
+                            fallback_policy = explicit_policy.delete(:fallback_policy)
+                            h[port_pair] = policy_for(
+                                source_task, *port_pair, sink_task,
+                                fallback_policy, explicit_policy: explicit_policy
```
`explicit_policy` is a keyword argument because some external tests call `policy_for`, and having it as a positional argument would require changes to those tests as well.

jhonasiv commented on lib/syskit/network_generation/dataflow_dynamics.rb:
```diff
@@ -618,9 +627,27 @@ def policy_for(
+
+            def policy_for(
+                source_task, source_port_name, sink_port_name, sink_task,
+                fallback_policy, explicit_policy: nil
+            )
+                computed_policy = {}
+                explicit_policy = {} unless explicit_policy
```
```suggestion
                fallback_policy, explicit_policy: {}
            )
                computed_policy = {}
```

You mentioned making this a kw arg due to not wanting to fix the policy_for call alls over the code, but you can make this a positional argument with a default value as well, that would work just like the kw arg worked.

doudou replied:
This. I think that's the last change.

doudou commented:
@jhonasiv updated, please have a look

eduardacoppo committed:
chore: use positional arg with default value

doudou reviewed (commented):
Adding empty review to get it out of my queue (since I'm the one working on it now)

doudou committed:
chore: minor cleanup for clarity

Removed the one-line, one-use policy_default_init_flag method,
and changed the argument order of merge_policy to match the hash
merge (i.e. computed, explicit instead of the opposite)
